File: src/analysis/identity.js
'use strict'

const crypto = require('crypto')
const { PERSISTENCE } = require('../config')

// BLE addresses rotate. Tracking a device by MAC alone means a tracker that
// changes address every 15 minutes looks like a stream of unrelated strangers,
// and the "seen for 30 minutes" test never fires.
//
// The fix is to fingerprint the parts of an advertisement that do *not* rotate
// and use timing to link an address that goes quiet to one that appears.
//
// For Apple Find My specifically, everything after the first four bytes of
// manufacturer data is the rotating public key, so only the company ID, type
// and length are usable. That is coarse — two AirTags fingerprint identically —
// which is why a link is only claimed when exactly one candidate fits.
function fingerprint(device) {
  const parts = []

  const mfr = device.manufacturerHex
  if (mfr && mfr.length >= 8) {
    // Company ID + type + length. Stable across key rotation.
    parts.push(`m:${mfr.slice(0, 8)}`)
    parts.push(`ml:${mfr.length}`)
  } else if (mfr) {
    parts.push(`m:${mfr}`)
  }

  const uuids = (device.serviceUuids || [])
    .map(u => String(u).toLowerCase().replace(/-/g, ''))
    .sort()
  if (uuids.length) parts.push(`s:${uuids.join(',')}`)

  if (device.name && device.name !== 'Unknown Device') parts.push(`n:${device.name}`)
  if (device.txPowerLevel != null) parts.push(`t:${device.txPowerLevel}`)

  if (parts.length === 0) return null

  return crypto.createHash('sha1').update(parts.join('|')).digest('hex').slice(0, 16)
}

class IdentityResolver {
  constructor(opts = {}) {
    this._identities = new Map()
    this._byAddress = new Map()
    this._byFingerprint = new Map()
    this._nextId = 1
    this._gapMs = opts.macRotationGapMs || PERSISTENCE.macRotationGapMs
    this._rssiDelta = opts.macRotationRssiDelta || PERSISTENCE.macRotationRssiDelta
    // An address still advertising is not an address that rotated away.
    this._quietMs = opts.quietMs || 10000
  }

  resolve(device) {
    const now = Date.now()
    const address = device.address
    const fp = fingerprint(device)

    const existingId = this._byAddress.get(address)
    if (existingId) {
      const identity = this._identities.get(existingId)
      identity.lastSeen = now
      identity.lastRssi = device.rssi
      const addr = identity.addresses.get(address)
      if (addr) { addr.lastSeen = now; addr.rssi = device.rssi }
      return this._view(identity, false)
    }

    if (fp) {
      const candidates = []
      for (const id of this._byFingerprint.get(fp) || []) {
        const identity = this._identities.get(id)
        if (!identity) continue
        if (identity.addresses.has(address)) continue

        const quietFor = now - identity.lastSeen
        if (quietFor < this._quietMs) continue          // still talking: different device
        if (quietFor > this._gapMs) continue            // too long ago to link

        if (
          identity.lastRssi != null && device.rssi != null &&
          Math.abs(identity.lastRssi - device.rssi) > this._rssiDelta
        ) continue                                      // different distance: different device

        candidates.push(identity)
      }

      // Ambiguity is not evidence. Two identical trackers in range must stay
      // two identities rather than be silently merged into one.
      if (candidates.length === 1) {
        const identity = candidates[0]
        const previousAddress = identity.currentAddress
        identity.addresses.set(address, { firstSeen: now, lastSeen: now, rssi: device.rssi })
        identity.currentAddress = address
        identity.lastSeen = now
        identity.lastRssi = device.rssi
        identity.rotations.push({ from: previousAddress, to: address, ts: now })
        this._byAddress.set(address, identity.id)
        return this._view(identity, true)
      }
    }

    const id = `id${this._nextId++}`
    const identity = {
      id,
      fingerprint: fp,
      currentAddress: address,
      addresses: new Map([[address, { firstSeen: now, lastSeen: now, rssi: device.rssi }]]),
      firstSeen: now,
      lastSeen: now,
      lastRssi: device.rssi,
      rotations: [],
    }
    this._identities.set(id, identity)
    this._byAddress.set(address, id)
    if (fp) {
      if (!this._byFingerprint.has(fp)) this._byFingerprint.set(fp, new Set())
      this._byFingerprint.get(fp).add(id)
    }
    return this._view(identity, false)
  }

  _view(identity, rotated) {
    return {
      identityId: identity.id,
      fingerprint: identity.fingerprint,
      rotated,
      addressCount: identity.addresses.size,
      addresses: Array.from(identity.addresses.keys()),
      previousAddress: rotated
        ? identity.rotations[identity.rotations.length - 1].from
        : null,
      rotations: identity.rotations.length,
      firstSeen: identity.firstSeen,
    }
  }

  get(id) {
    return this._identities.get(id) || null
  }

  clearOld(maxAgeMs) {
    const cutoff = Date.now() - maxAgeMs
    for (const [id, identity] of this._identities) {
      if (identity.lastSeen >= cutoff) continue
      this._identities.delete(id)
      for (const addr of identity.addresses.keys()) this._byAddress.delete(addr)
      if (identity.fingerprint) {
        const set = this._byFingerprint.get(identity.fingerprint)
        if (set) {
          set.delete(id)
          if (set.size === 0) this._byFingerprint.delete(identity.fingerprint)
        }
      }
    }
  }
}

module.exports = { IdentityResolver, fingerprint }

File: android/app/src/main/java/com/trackdetect/analysis/Identity.kt
package com.trackdetect.analysis

import android.bluetooth.le.ScanRecord
import java.security.MessageDigest
import kotlin.math.abs

/**
 * BLE addresses rotate. Tracking by MAC alone means a tracker that changes
 * address every fifteen minutes looks like a stream of unrelated strangers, and
 * the "present for thirty minutes" test never fires.
 *
 * So devices are fingerprinted on the parts of an advertisement that do not
 * rotate. For Apple Find My specifically, everything after the first two bytes
 * of the payload is the rotating public key, leaving only the type and length —
 * coarse enough that two AirTags fingerprint identically, which is exactly why
 * a rotation is only linked when there is precisely one candidate.
 */
object Fingerprint {

    fun of(record: ScanRecord?): String? {
        if (record == null) return null
        val parts = ArrayList<String>()

        val mfr = record.manufacturerSpecificData
        if (mfr != null && mfr.size() > 0) {
            for (i in 0 until mfr.size()) {
                val company = mfr.keyAt(i)
                val payload = mfr.valueAt(i) ?: continue
                val head = payload.take(2).joinToString("") { "%02x".format(it) }
                parts.add("m:%04x:%s:%d".format(company, head, payload.size))
            }
        }

        record.serviceUuids
            ?.map { it.uuid.toString() }
            ?.sorted()
            ?.takeIf { it.isNotEmpty() }
            ?.let { parts.add("s:" + it.joinToString(",")) }

        record.deviceName?.takeIf { it.isNotBlank() }?.let { parts.add("n:$it") }

        val tx = record.txPowerLevel
        if (tx != Int.MIN_VALUE) parts.add("t:$tx")

        if (parts.isEmpty()) return null

        val digest = MessageDigest.getInstance("SHA-1").digest(parts.joinToString("|").toByteArray())
        return digest.take(8).joinToString("") { "%02x".format(it) }
    }
}

class IdentityResolver(
    private val rotationWindowMs: Long = 3 * 60 * 1000,
    private val quietMs: Long = 10_000,
    private val rssiDelta: Int = 12
) {

    class Identity(val id: String, val fingerprint: String?) {
        val addresses = LinkedHashSet<String>()
        var currentAddress: String = ""
        var lastSeen: Long = 0
        var lastRssi: Int = 0
        var rotations: Int = 0
    }

    data class Resolution(
        val identity: Identity,
        val rotated: Boolean,
        val previousAddress: String?
    )

    private val byId = HashMap<String, Identity>()
    private val byAddress = HashMap<String, String>()
    private val byFingerprint = HashMap<String, MutableSet<String>>()
    private var counter = 0

    // resolve() runs on the Bluetooth callback thread and prune() on the service's
    // coroutine, so every access to the three maps below has to be guarded. Without
    // this the pruning pass iterates a map the scan thread is writing to and the
    // service dies on ConcurrentModificationException mid-scan.
    @Synchronized
    fun prune(now: Long, maxAgeMs: Long = 30 * 60_000L) {
        val cutoff = now - maxAgeMs
        byId.values.filter { it.lastSeen < cutoff }.map { it.id }.forEach { forget(it) }

        // A thirty-minute window in a station or a shopping centre, with every
        // privacy-conscious device rotating its address every quarter hour, is enough
        // to accumulate a very large number of one-off identities. Cap it by dropping
        // the least recently heard — they are the ones least likely to be on you.
        if (byId.size > MAX_IDENTITIES) {
            byId.values
                .sortedBy { it.lastSeen }
                .take(byId.size - MAX_IDENTITIES)
                .map { it.id }
                .forEach { forget(it) }
        }
    }

    private fun forget(id: String) {
        val identity = byId.remove(id) ?: return
        identity.addresses.forEach { byAddress.remove(it) }
        if (identity.fingerprint != null) {
            byFingerprint[identity.fingerprint]?.let { set ->
                set.remove(id)
                if (set.isEmpty()) byFingerprint.remove(identity.fingerprint)
            }
        }
    }

    /** Forgets every identity. Used by "clear all data". */
    @Synchronized
    fun clear() {
        byId.clear()
        byAddress.clear()
        byFingerprint.clear()
    }

    @Synchronized
    fun resolve(address: String, rssi: Int, fingerprint: String?, now: Long): Resolution {
        // byId may not hold the id if a prune raced ahead of a stale byAddress entry;
        // fall through to re-creating the identity rather than throwing.
        byAddress[address]?.let { id -> byId[id] }?.let { identity ->
            identity.lastSeen = now
            identity.lastRssi = rssi
            return Resolution(identity, false, null)
        }

        if (fingerprint != null) {
            val candidates = byFingerprint[fingerprint]
                ?.mapNotNull { byId[it] }
                ?.filter { candidate ->
                    if (candidate.addresses.contains(address)) return@filter false
                    val quietFor = now - candidate.lastSeen
                    // Still advertising means it is a different device, not a rotation.
                    if (quietFor < quietMs) return@filter false
                    if (quietFor > rotationWindowMs) return@filter false
                    abs(candidate.lastRssi - rssi) <= rssiDelta
                }
                ?: emptyList()

            // Ambiguity is not evidence. Two identical trackers in range stay two
            // identities rather than being silently merged into one.
            if (candidates.size == 1) {
                val identity = candidates[0]
                val previous = identity.currentAddress
                identity.addresses.add(address)
                identity.currentAddress = address
                identity.lastSeen = now
                identity.lastRssi = rssi
                identity.rotations++
                byAddress[address] = identity.id
                return Resolution(identity, true, previous)
            }
        }

        val identity = Identity("id${++counter}", fingerprint)
        identity.addresses.add(address)
        identity.currentAddress = address
        identity.lastSeen = now
        identity.lastRssi = rssi
        byId[identity.id] = identity
        byAddress[address] = identity.id
        if (fingerprint != null) {
            byFingerprint.getOrPut(fingerprint) { HashSet() }.add(identity.id)
        }
        return Resolution(identity, false, null)
    }

    private companion object {
        /** Ceiling on simultaneously remembered identities — see [prune]. */
        const val MAX_IDENTITIES = 4000
    }
}
